### PR TITLE
re-add support for jsonpretty + yaml to image based gadgets

### PR DESCRIPTION
**Pull request by flyth**
This adds the missing `jsonpretty` and `yaml` output modes to the cli operator. Currently (as was with the previous
run gadget), YAML is handled by first marshaling to JSON and then converting, which is quite slow. We can later
on add an optimized YAML formatter that will handle marshaling directly.

This also sorts fields alphabetically.

```
$ ./ig run "ghcr.io/inspektor-gadget/gadget/trace_exec:latest" -o jsonpretty
INFO[0000] Experimental features enabled
{
  "args": "/opt/cni/bin/bridge",
  "args_count": 1,
  "args_size": 20,
  "comm": "bridge",
  "gid": 0,
  "k8s": {
    "container": "",
    "hostnetwork": false,
    "namespace": "",
    "node": "",
    "pod": ""
  },
  "loginuid": 4294967295,
  "mntns_id": 4026532263,
  "pid": 1475846,
  "ppid": 8590,
  "retval": 0,
  "runtime": {
    "containerId": "e127eea07a1ecd3d8a75515486c31fa6f685fd0ad7d98a598ac3ea70db0d4ed4",
    "containerImageDigest": "",
    "containerImageName": "gcr.io/k8s-minikube/kicbase:v0.0.40",
    "containerName": "minikube",
    "runtimeName": "docker"
  },
  "sessionid": 4294967295,
  "timestamp": "2024-03-27T23:41:35.264215823+01:00",
  "uid": 0,
  "upper_layer": false
}
...
```

### CODE REVIEW

File: pkg/operators/cli/clioperator.go [hunk 1]
```diff
@@ -241,15 +245,34 @@ func (o *cliOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error
...
+
+			if o.mode == ModeYAML {
+				// For the time being, this uses a slow approach to marshal to YAML, by first
+				// converting to JSON and then to YAML. This should get a dedicated formatter sooner or later.
+				df = func(ds datasource.DataSource, data datasource.Data) error {
+					yml, err := yaml.JSONToYAML(jsonFormatter.Marshal(data))
+					if err != nil {
+						return fmt.Errorf("serializing yaml: %w", err)
+					}
+					fmt.Println("---")
```
Comment (blanquicet) [inline pkg/operators/cli/clioperator.go:270]: It's not strictly related to this PR, but just want to highlight that the CLI operator still needs to start using the `frontends/console` pkg.

File: pkg/operators/cli/clioperator.go [hunk 2]
```diff
@@ -241,15 +245,34 @@ func (o *cliOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error
...
+					fmt.Println(string(yml))
```
Comment (blanquicet) [inline pkg/operators/cli/clioperator.go:271]: `yml` already contains a final newline:

```suggestion
					fmt.Print(string(yml))
```

Current output:

```bash
INFO[0000] Experimental features enabled
---
ino: 39847
k8s:
  container: ""
  hostnetwork: false
  namespace: ""
  node: ""
  pod: ""
netns: 4026532260
runtime:
  containerId: 7bae168baa38838a73e1a4b384c526af717cfd2b45197bd2833bf4863466cc9d
  containerImageDigest: ""
  containerImageName: nginx:1.21
  containerName: shell01
  runtimeName: docker
state: 10

---
ino: 39846
k8s:
  container: ""
  hostnetwork: false
  namespace: ""
  node: ""
  pod: ""
netns: 4026532260
runtime:
  containerId: 7bae168baa38838a73e1a4b384c526af717cfd2b45197bd2833bf4863466cc9d
  containerImageDigest: ""
  containerImageName: nginx:1.21
  containerName: shell01
  runtimeName: docker
state: 10

```

After this change:

```bash
INFO[0000] Experimental features enabled
---
ino: 39847
k8s:
  container: ""
  hostnetwork: false
  namespace: ""
  node: ""
  pod: ""
netns: 4026532260
runtime:
  containerId: 7bae168baa38838a73e1a4b384c526af717cfd2b45197bd2833bf4863466cc9d
  containerImageDigest: ""
  containerImageName: nginx:1.21
  containerName: shell01
  runtimeName: docker
state: 10
---
ino: 39846
k8s:
  container: ""
  hostnetwork: false
  namespace: ""
  node: ""
  pod: ""
netns: 4026532260
runtime:
  containerId: 7bae168baa38838a73e1a4b384c526af717cfd2b45197bd2833bf4863466cc9d
  containerImageDigest: ""
  containerImageName: nginx:1.21
  containerName: shell01
  runtimeName: docker
state: 10
```

File: pkg/datasource/formatters/json/json.go [hunk 3]
```diff
@@ -115,6 +116,12 @@ func (f *Formatter) addSubFields(accessors []datasource.FieldAccessor, prefix st
 	}
 
 	ctr := -1
+
+	// sort alphabetically
+	sort.Slice(accessors, func(i, j int) bool {
+		return accessors[i].Name() < accessors[j].Name()
+	})
```
Comment (blanquicet) [inline pkg/datasource/formatters/json/json.go:123]: I like this change, but I also like the idea of having things in this order:

1. Runtime-specific stuff
2. Kubernetes-specific stuff
3. Gadget-specific stuff

Where 1 and 2 can be exchanged, it's also fine.

I don't know, what do others think?

File: pkg/datasource/formatters/json/options.go [hunk 4]
```diff
@@ -37,3 +37,10 @@ func WithShowAll(val bool) Option {
 		formatter.useDefault = true
 	}
 }
+
+func WithPretty(val bool, indent string) Option {
```
[the remark this answers is not in the document]
Comment (flyth) [reply pkg/datasource/formatters/json/options.go:41]: I prefer to use explicit values, as it allows avoiding additional conditionals if dealing with bool values on the caller side in the first place, e.g.:

```golang
isPretty := true
...
json.New(json.WithPretty(isPretty, "  "), json.WithMaybeOtherOption())
```

is easier than

```golang
isPretty := true
...
options := []json.Option{json.WithMaybeOtherOption()}
if isPretty {
    options = append(options, json.WithPretty("  "))
}
json.New(json.WithPretty(isPretty, "  "), json.WithMaybeOtherOption())
```

Just looked at the code again - let's practice what we preach and do it ourselves as well 🙈 :

```golang
jsonFormatter, err := json.New(ds,
    json.WithShowAll(true),
    json.WithPretty(o.mode == ModeJSONPretty, "  "),
)
```

File: pkg/datasource/formatters/json/json.go (hunk 3 above)
Comment (flyth) [reply pkg/datasource/formatters/json/json.go]: Main reason was: the Go JSON marshaler sorts by keys as well. As a human I also find it easier to navigate :)

File: pkg/datasource/formatters/json/options.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/datasource/formatters/json/options.go:41]: Ok, makes sense.

File: pkg/datasource/formatters/json/json.go (hunk 3 above)
Comment (blanquicet) [reply pkg/datasource/formatters/json/json.go]: Easier to navigate but, IMO, not very good looking. Anyway, given that no one else commented, let's go ahead. Once it's merged and people will start using it, we will realize if people like it or not.

**Pull request merged**